pulp_container/app/models.py

```diff
@@ -452,6 +460,71 @@ class Meta:
         default_related_name = "%(app_label)s_%(model_name)s"
 
 
+class ManifestSigningService(SigningService):
+    """
+    Signing service used for creating container signatures.
+    """
+
+    def validate(self):
+        """
+        Validate a signing service for a container signature.
+
+        The validation seeks to ensure that the sign() method returns a dict as follows:
+
+        {"signature": "$SIG_PATH"}
+
+        The method creates a test image manifest, signs its manifest.json, and checks if the
+        signature can be verified by the provided public key.
+
+        Raises:
+            RuntimeError: If the validation has failed.
+
+        """
+        test_manifest = {
+            "schemaVersion": 2,
+            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+            "config": {
+                "mediaType": "application/vnd.docker.container.image.v1+json",
+                "size": 1456,
+                "digest": "sha256:7138284460ffa3bb6ee087344f5b051468b3f8697e2d1427bac1a208d4168123",
+            },
+            "layers": [
+                {
+                    "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
+                    "size": 772792,
+                    "digest": "sha256:e685c5c858e36338a47c627763b50dfe6035b547f1f75f0d39753d4e3121",
+                }
+            ],
+        }
+
+        with tempfile.TemporaryDirectory() as temp_directory_name:
```

Comment:
since this is not a task, it will not use tasks' directory but the one from which the signing service command was called. it will fail like this because it would try to remove all files while traversing the tree and hit permission issues.
``
  File "/usr/lib64/python3.9/tempfile.py", line 826, in __exit__
    self.cleanup()
  File "/usr/lib64/python3.9/tempfile.py", line 830, in cleanup
    self._rmtree(self.name)
  File "/usr/lib64/python3.9/tempfile.py", line 812, in _rmtree
    _shutil.rmtree(name, onerror=onerror)
  File "/usr/lib64/python3.9/shutil.py", line 730, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/usr/lib64/python3.9/shutil.py", line 728, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: './tmpa8wmdpgv'
``

Reply:
Are you sure it uses the directory from which the signing service command was called? According to the [docs](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp) (it took me there from [`TemporaryDirectory`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory)), `the default directory is chosen from a platform-dependent list, but the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables.`

Reply:
> since this is not a task, it will not use tasks' directory but the one from which the signing service command was called. it will fail like this because it would try to remove all files while traversing the tree and hit permission issues. `File "/usr/lib64/python3.9/tempfile.py", line 826, in __exit__ self.cleanup() File "/usr/lib64/python3.9/tempfile.py", line 830, in cleanup self._rmtree(self.name) File "/usr/lib64/python3.9/tempfile.py", line 812, in _rmtree _shutil.rmtree(name, onerror=onerror) File "/usr/lib64/python3.9/shutil.py", line 730, in rmtree onerror(os.rmdir, path, sys.exc_info()) File "/usr/lib64/python3.9/shutil.py", line 728, in rmtree os.rmdir(path) OSError: [Errno 39] Directory not empty: './tmpa8wmdpgv'`

Ouch.
I wonder if it's a good topic to discuss with a wider group. If we need to write a small file and it's outside of the task, what's the best approach? 
I'd use pulp temp directory.